crates/lace-bin/src/typechecking/constraint_generation.rs

```diff
@@ -0,0 +1,338 @@
+                        ctx.concrete_type(ConcreteType::Int)
+                    }
+                }
+            }
+            ExprKind::Tuple(t) => {
+                let mut res = Vec::new_in(ctx.arena);
+                for i in *t {
+                    let ty = i.generate_constraints(ctx);
+                    res.push(ty);
+                }
+
+                ctx.concrete_type(ConcreteType::Tuple(res.into_bump_slice()))
+            }
+            ExprKind::Call(f, args) => {
+                let mut param_types = Vec::new_in(ctx.arena);
+                for i in args.value {
+                    param_types.push(i.generate_constraints(ctx));
+                }
+
+                let ret_ty = ctx.fresh();
+
+                let expected_f_ty = ConcreteType::Function {
+                    params: param_types.into_bump_slice(),
+                    ret: ctx.alloc(ret_ty),
+                };
+                let defined_f_ty = f.generate_constraints(ctx);
+
+                ctx.add_equal_constraint_concrete(
+                    defined_f_ty,
+                    expected_f_ty,
+                    ConstraintMetadata::Call {
+                        call_expr: self.node_id,
+                    },
+                );
+
+                ret_ty
+            }
+        }
+    }
+}
+
+/// The types of ways in which a block could have returns in it
+/// TODO: explicit return
+pub enum BlockReturn {
+    /// last item has semi, unit return
+    None,
+    /// return expression
+    Implicit(NodeId),
+}
+
+impl<'a, 'sp> TypeConstraintGenerator<'a, 'sp> for Identified<Block<'_, '_>> {
+    type TypeResult = (TypeVariable, BlockReturn);
+
+    fn generate_constraints(&self, ctx: &mut TypeContext<'a, 'sp>) -> Self::TypeResult {
+        for i in self.value.stmts {
+            i.generate_constraints(ctx);
+        }
+
+        if let Some(ref last) = self.value.last {
+            (
+                last.generate_constraints(ctx),
+                BlockReturn::Implicit(last.node_id),
+            )
+        } else {
+            (ctx.concrete_type(ConcreteType::Unit), BlockReturn::None)
+        }
+    }
+}
+
+impl<'a, 'sp> TypeConstraintGenerator<'a, 'sp> for Lit<'_> {
+    type TypeResult = TypeVariable;
+
+    fn generate_constraints(&self, ctx: &mut TypeContext<'a, 'sp>) -> Self::TypeResult {
+        match self {
+            Lit::Bool(_) => ctx.concrete_type(ConcreteType::Bool),
+            Lit::Int(_) => ctx.concrete_type(ConcreteType::Int),
+            Lit::String(_) => ctx.concrete_type(ConcreteType::String),
+        }
+    }
+}
+
+impl<'a, 'sp> TypeConstraintGenerator<'a, 'sp> for Statement<'_, '_> {
+    type TypeResult = ();
+
+    fn generate_constraints(&self, ctx: &mut TypeContext<'a, 'sp>) -> Self::TypeResult {
+        match self {
+            Statement::Expr(e) => {
+                // whatever :shrug:
+                let _ = e.generate_constraints(ctx);
```

Comment:
Why do we not care?

Love the comment btw

Reply:
because the outcome type of an expression in statement position doesn't matter. The type of `3 + 3;` is int. But we don't care, we don't use that outcome.

Reply:
Ah so there's no Result were ignoring on anything like that? All good then!